tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+            out: self.out.clone(),
+        }
+    }
+}
+
+impl<W> Drop for FlameGuard<W>
+where
+    W: Write + 'static,
+{
+    fn drop(&mut self) {
+        dbg!(());
+        match self.out.lock().unwrap().flush() {
+            Ok(_) => (),
+            Err(e) => eprintln!("Error: {}", e),
+        }
+    }
+}
+
+impl<S> FlameLayer<S, BufWriter<File>>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    pub fn write_to_file(path: impl AsRef<Path>) -> Result<Self, Error> {
+        let path = path.as_ref();
+        let file = File::create(path).map_err(|source| Error::IO {
+            path: path.into(),
+            source,
+        })?;
+        let writer = BufWriter::new(file);
+        Ok(Self::new(writer))
+    }
+}
+
+impl<S, W> Layer<S> for FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    fn new_span(&self, _: &Attributes, id: &Id, ctx: Context<S>) {
+        let samples = self.time_since_last_event();
+
+        let parent = ctx.span(id).expect("expected: span id exists in registry").parent();
+
+        let mut stack = String::new();
+
+        stack.push_str("tracing");
+
+        if let Some(parent) = parent {
+            for parent in parent.from_root() {
+                stack.push_str("; ");
+                write(&mut stack, parent).expect("expected: write to String never fails");
+            }
+        }
+
+        write!(&mut stack, " {}", samples.as_nanos())
+            .expect("expected: write to String never fails");
+        writeln!(
+            *self.out.lock().unwrap(),
+            "{}",
+            stack
+        )
+        .expect("expected: write to String never fails");
+    }
+
+    #[allow(clippy::needless_return)]
+    fn on_close(&self, id: Id, ctx: Context<S>) {
+        let panicking = std::thread::panicking();
+        macro_rules! expect {
+            ($e:expr, $msg:literal) => {
+                if panicking {
+                    return;
+                } else {
+                    $e.expect($msg)
+                }
+            };
+        }
+
+        let samples = self.time_since_last_event();
+        let first = expect!(ctx.span(&id), "expected: span id exists in registry");
+        let parent = first.parent();
+
+        let mut stack = String::new();
+        stack.push_str("tracing; ");
+
+        if let Some(parent) = parent {
+            for parent in parent.from_root() {
+                expect!(write(&mut stack, parent), "expected: write to String never fails");
+                stack.push_str("; ");
+            }
+        }
+
+        expect!(write(&mut stack, first), "expected: write to String never fails");
+        expect!(write!(&mut stack, " {}", samples.as_nanos()), "expected: write to String never fails");
+        expect!(writeln!(
+            *expect!(self.out.lock(), "expected: lock is never poisoned"),
+            "{}",
+            stack
+        ), "expected: write to String never fails");
```

Comment:
this last `writeln!` is writing to an IO, not to a string. is it correct to panic if this fails?